### ISSUE TITLE
[Feature Request] Creating list of releases on label's page

### ISSUE THREAD

**Issue by viviancels** [DJDoubleD/QobuzDownloaderX-MOD#106]
# **🚀 Feature Request**

## **Is your feature request related to a problem? Please describe.**

Hello. Recently, the site introduced restrictions on the number of releases in the list on the label page. Example:
https://www.qobuz.com/**se-en**/label/the-dance-division/download-streaming-albums/287307
If you copy-past the link and download it using your Swedish account, you will receive 1406 releases
Byt some releases did not download (I downloaded manualy, also inserting links with "se-en" and using sweden accaount ):
https://www.qobuz.com/se-en/album/gonna-be-alright-gregoir-cruz/hlt9f1gly5l5a
https://www.qobuz.com/se-en/album/scooby-doo-pa-pa-don-lore-v/duiggmbjp9nrb
https://www.qobuz.com/se-en/album/memories-dj-jump/a28ys78et29ec
https://www.qobuz.com/se-en/album/secrets-kery-fay/hdpyntt15gp5a
For some reason, these releases did not download automatically when downloading the label
...
Another example 1000 releases limit:
https://www.qobuz.com/se-en/label/armada-music/download-streaming-albums/248166
In search shows 5416 releases on Label, in fact, there were much fewer releases in the summer
---
I always used to create a list of releases manually using regular expressions in source code page:
`.*?<div class="product__item">.*?href="(.*?)" .*?tooltip">(.*?)<.*?<p class="product__artist">.*?">(.*?)<\/a>.*?|.*<\/html>
$3	$2	https://www.qobuz.com$1\n`

But now, due to the limitation of the number of releases on label's page, this makes no sense. However, QobuzDownloaderX-MOD somehow learns about the rest of the releases that are not visible on the label's page (even if it misses some releases).

## **Describe the solution you'd like**

Creating a list (album artist - album title (full date) link) would allow you to find out which releases are available

---

## **Describe alternatives you've considered**

*

---

### **Additional context**

*

**Comment by DJDoubleD** [comment]
There is a difference between the main qobuz.com website, which is basically a storefront and the Qobuz streaming platforms (web app and mobile apps).
I don't know what the main site uses to fetch the max 1000 albums from a label (I'm guessing they are using the `catalog/search` endpoint as it seems to be limited to 1000 results as well).

In any case, QobuzDownloaderX-MOD uses the same API as the web player's label page, which is still limited to 10000 results (see #101). I can only assume the results from that API are accurate and have no way of automatically comparing them to what the main site shows. 

Implementing a list feature in the application would be theoretically possible, but there is no way that winforms will be able to display thousands of results in a performant way (at least that I know of).
Rendering performance is the reason why I limit the search results to 15 items.

In any case, as there is only 1 API endpoint that's used to retrieve the albums from a label, it is fairly simple to emulate what the application does to retrieve the albums for download.
As an example, I've attached a [small python script](https://github.com/user-attachments/files/17266569/qobuz_get_label_albums.zip) you can use to generate a json file with all (up to 10000) albums from a label. (had to zip it because github doesn't allow .py files, apparently)
The output from this list is what QobuzDownloaderX-MOD will use when you download a label.

Maybe this will help you in finding the discrepancies and get the albums that are on the main site, but not in the downloaded items.

**Comment by viviancels** [comment]
@DJDoubleD  thank you. The script requests app_id, authorization token and label id

**Comment by DJDoubleD** [comment]
> @DJDoubleD thank you. The script requests app_id, authorization token and label id

@viviancels Yes, I made those input variables deliberate as I try not to publish the app_id on GitHub.
You can get the app_id using [this method](https://gist.github.com/vitiko98/bb89fd203d08e285d06abf40d96db592) (I just tried it and still works). Normally it doesn't change, so you can hard code it in the script yourself if you want.

The authorization token is just the token for the account you use, see [my wiki](https://github.com/DJDoubleD/QobuzDownloaderX-MOD/wiki/Logging-In-(The-Alternate-Way)) for instructions if you need to get it from your personal account.
The label id is in the label url that you use to download the label.
So in your example `https://www.qobuz.com/se-en/label/the-dance-division/download-streaming-albums/287307`, the label id is `287307`.

**Comment by viviancels** [comment]
@DJDoubleD thank you. It works. I saw the json code and  noticed that albums (releases) have two identifiers:
1.       "url": "/se-en/album/-/m31zcug4bfd3b",
2.       "id": "m31zcug4bfd3b",
I believe that QobuzDownloaderX-MOD uses the second identifier and adds the beginning of the address to it https://www.qobuz.com/fr-fr/album/. Perhaps it would be more accurate to use the first identifier (/se-en/album/-/m31zcug4bfd3b), due to the fact that when a release has territorial restrictions, the repository page cannot be opened if its address contains the code of another country (se-en). The first release identifier contains the country code specified in the label reference. I am writing this here so as not to create a new topic.

**Comment by DJDoubleD** [comment]
> I believe that QobuzDownloaderX-MOD uses the second identifier and adds the beginning of the address to it https://www.qobuz.com/fr-fr/album/. Perhaps it would be more accurate to use the first identifier (/se-en/album/-/m31zcug4bfd3b), due to the fact that when a release has territorial restrictions, the repository page cannot be opened if its address (link) contains the code of another country (se-en). The first release identifier contains the country code specified in the label reference. I am writing this here so as not to create a new topic.

Just to avoid confusion, QobuzDownloaderX-MOD does not generate links like `https://www.qobuz.com/fr-fr/album/...` to get the album details.
It uses the Qobuz API directly ([via the API library I made](https://github.com/DJDoubleD/QobuzApiSharp/blob/c227398a6281a55365bb7425992de5d245b3de94/QobuzApiSharp/Api/Service/Endpoints/QobuzApiService.Album.cs#L23)) to get the album data.
Just like in the script I added to this issue, it's the user_auth_token that determines the region (region of the account) and only the albums that are available for that region will be returned and attempted to be downloaded.

There is no way to explain the discrepancies between the numbers that the main www.qobuz.com website gives and what the API returns. I can only work with what the API returns...

**Comment by ghost** [comment]
> In any case, as there is only 1 API endpoint that's used to retrieve the albums from a label, it is fairly simple to emulate what the application does to retrieve the albums for download. As an example, I've attached a [small python script](https://github.com/user-attachments/files/17266569/qobuz_get_label_albums.zip) you can use to generate a json file with all (up to 10000) albums from a label. (had to zip it because github doesn't allow .py files, apparently) The output from this list is what QobuzDownloaderX-MOD will use when you download a label.

Thank you so so much for this :) so kind of you to write it and share. It really helps me out, and so does the rest of the info in your comments as someone trying to learn :) have a lovely day

**Comment by orbittwz** [comment]
should be closed.
